packages/warning/src/index.js

```diff
@@ -0,0 +1,45 @@
+function isDev() {
+	return (
+		typeof process !== 'undefined' &&
+		process.env &&
+		process.env.NODE_ENV !== 'production'
+	);
+}
+
+/**
+ * Shows a warning with `messages` if `condition` passes and environment is not `production`.
+ *
+ * @param {boolean} condition Whether the warning will be triggered or not.
+ * @param {string[]} messages Message(s) to show in the warning.
+ *
+ * @example
+ * ```js
+ * import warning from '@wordpress/warning';
+ *
+ * function MyComponent( props ) {
+ *   warning( ! props.title, '`props.title` was not passed' );
+ *   ...
+ * }
+ * ```
+ */
+export default function warning( condition, ...messages ) {
+	if ( isDev() ) {
+		if ( ! condition ) {
+			return;
+		}
+
+		const text = messages.join( '\n' );
```

Comment:
Not really concerned too much over this, but was there any prior art for choosing `\n` here?

Notably, if we had implemented this as something like:

```js
console.warn( ...messages );
```

`warn` would join _using a space_.

Technically speaking, for `console` logging, this behavior is not specified, and left up to the implementation, so a newline could be reasonable as well:

>How the implementation prints args is up to the implementation, but implementations should separate the objects by a space or something similar, as that has become a developer expectation.

https://console.spec.whatwg.org/#printer

Reply:
This also raises a question: Why do we support multiple `message` strings? Is there a demonstrated need for it? Would it just be simpler to accept only a single string, and consider to add support for the variadic arguments if it becomes a problem? Especially considering the specific treatment of `messages` is not currently documented very thoroughly, it seems like a feature that [we might not need](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Reply:
I tried to mimic the same API as the `warning` function on Reakit, [which uses the multiline feature a lot](https://github.com/reakit/reakit/search?q=warning&unscoped_q=warning). But I completely agree that it's better to start with a single parameter for now.

Reply:
Thanks for linking to the prior art. That definitely helps clarify the way in which you were expecting it to be used. At least for some of the examples I see there, it reminds me a bit of this project's [`@wordpress/deprecated` module](https://github.com/WordPress/gutenberg/tree/master/packages/deprecated), at least so far as having some consistency in how those messages are structured (a "title" line, a "see more" line, etc). The API for `deprecated` is not always obvious, and I can see merit in having the low-level APIs like `warning` be unopinionated about this sort of thing. But I do think it's something in the future where if we start to see these sorts of patterns emerging, we can decide then whether it's something we want to leave non-formalized (and optimize for multi-line messages) or build other abstractions around (something more like `deprecated` which manages how the message is structured in a consistent way).